File: src/source/TelegramSource.ts
// src/TelegramSource.ts
import {TelegramClient} from 'telegram';
import {StringSession} from 'telegram/sessions';
import fs from 'fs';
import path from 'path';
import {Source} from './Source';
import {Channel, PrismaClient} from "@prisma/client";

export class TelegramSource implements Source {
    private client: TelegramClient;
    private channelUsername: string;
    private lastMessageIdFile: string;
    private dbClient: PrismaClient;

    constructor(apiId: number, apiHash: string, sessionFile: string, dbClient: PrismaClient, channelUsername: string) {
        let sessionString = '';
        this.dbClient = dbClient;
        // Check if session file exists and read it
        if (fs.existsSync(sessionFile)) {
            sessionString = fs.readFileSync(sessionFile, 'utf-8');
        } else {
            console.error(`Session file not found at ${sessionFile}. Make sure to run the session utility first.`);
            throw new Error('Session file not found.');
        }

        const stringSession = new StringSession(sessionString); // Load session from file
        this.client = new TelegramClient(stringSession, apiId, apiHash, {
            connectionRetries: 5,
        });
        this.channelUsername = channelUsername;
        this.lastMessageIdFile = path.join(__dirname, `${channelUsername}_lastMessageId.txt`);
    }

    async start() {
        await this.client.connect(); // Connect using the existing session
        console.log('Connected to Telegram using saved session.');
    }

    async fetchData(): Promise<string[]> {
        const channel = await this.client.getEntity(this.channelUsername);
        const lastMessageId = await this.getLastMessageId();
        let messages;
        if (lastMessageId == 0) {
            messages = await this.client.getMessages(channel, {limit: 10})
        }else{
            messages = await this.client.getMessages(channel, {limit: 10})
            // messages = await this.client.getMessages(channel, {
            //     minId: lastMessageId,
            // });
        }
        if (messages.length > 0) {
            await this.saveLastMessageId(messages[0].id);
        }

        return messages.map((message) => message.message);
    }

    async close() {
        await this.client.disconnect();
        console.log('Disconnected from Telegram');
    }

    private async getLastMessageId(): Promise<number> {
        const res =  await this.dbClient.channel.findUnique({
            where: {channelID: this.channelUsername},
        });
        return res?.messageID || 0;
    }

    private async saveLastMessageId(messageId: number) {
       await this.dbClient.channel.upsert({
            where: {channelID: this.channelUsername},
            update: {messageID: messageId},
            create: {channelID: this.channelUsername, messageID: messageId},
        });
    }
}

File: src/source/telegram_source.py
# src/source/telegram_source.py
from mpmath import limit

from src.logger_config import get_logger
from src.session_manager import SessionManager
from src.source.models import db, Channel
from src.settings import settings
from src.source.source import Source

logger = get_logger(__name__)

class TelegramSource(Source):
    def __init__(self):
        self.client = SessionManager().get_client()
        self.channels = settings.channels

        # Ensure tables are created
        db.connect()
        db.create_tables([Channel], safe=True)
        db.close()

    async def fetch_data(self) -> list:
        await self.client.connect()
        if not await self.client.is_user_authorized():
            logger.error("Client is not authorized. Please check your session string.")
            raise ValueError("Client is not authorized.")

        new_messages = []

        for channel_id in self.channels:
            last_message_id = self.get_last_message_id(channel_id)
            logger.debug(f"Fetching messages for channel {channel_id} starting from message ID {last_message_id}")

            try:
                # Fetch messages after last_message_id
                latest_message_id = last_message_id
                async for message in self.client.iter_messages(channel_id, min_id=last_message_id, limit=10):
                    if message.text:
                        reference = f"Channel: {channel_id}, Message ID: {message.id}"
                        new_messages.append((message.text, reference))
                        if message.id > latest_message_id:
                            latest_message_id = message.id

                if latest_message_id > last_message_id:
                    self.save_last_message_id(channel_id, latest_message_id)
            except Exception as e:
                logger.error(f"Error fetching messages from {channel_id}: {e}", exc_info=True)

        await self.client.disconnect()
        logger.debug("Disconnected Telegram client")

        return new_messages

    def get_last_message_id(self, channel_id: str) -> int:
        try:
            channel = Channel.get(Channel.channel_id == channel_id)
            last_id = channel.last_message_id or 0
        except Channel.DoesNotExist:
            last_id = 0
        logger.debug(f"Last message ID for channel {channel_id}: {last_id}")
        return last_id

    def save_last_message_id(self, channel_id: str, message_id: int):
        Channel.insert(channel_id=channel_id, last_message_id=message_id).on_conflict(
            update={Channel.last_message_id: message_id},
            conflict_target=[Channel.channel_id]
        ).execute()
        logger.debug(f"Saved last message ID {message_id} for channel {channel_id}")
